contracts/oracles/UniswapV2Oracle.sol

```diff
@@ -0,0 +1,276 @@
+
+import {ArrayHelper} from "../libs/arrays/ArrayHelper.sol";
+
+/**
+ * @notice UniswapV2Oracle module
+ *
+ * A contract for retrieving prices from Uniswap V2 pairs. Works by keeping track of pairs that were
+ * added as paths and returns prices of tokens following the configured routes.
+ *
+ * Arbitrary time window (time between oracle observations) may be configured and the Oracle will adjust automatically.
+ *
+ * From time to time `updatePrices()` function has to be called in order to calculate correct TWAP.
+ */
+abstract contract UniswapV2Oracle is Initializable {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using UniswapV2OracleLibrary for address;
+    using ArrayHelper for uint256[];
+    using Math for uint256;
+
+    struct PairInfo {
+        uint256[] prices0Cumulative;
+        uint256[] prices1Cumulative;
+        uint256[] blockTimestamps;
+        uint256 refs;
+    }
+
+    IUniswapV2Factory public uniswapV2Factory;
+
+    uint256 public timeWindow;
+
+    EnumerableSet.AddressSet private _pairs;
+    mapping(address => address[]) private _paths;
+    mapping(address => PairInfo) private _pairInfos;
+
+    /**
+     * @notice Constructor
+     * @param uniswapV2Factory_ the Uniswap V2 factory
+     * @param timeWindow_ the time between oracle observations
+     */
+    function __OracleV2_init(
+        address uniswapV2Factory_,
+        uint256 timeWindow_
+    ) internal onlyInitializing {
+        uniswapV2Factory = IUniswapV2Factory(uniswapV2Factory_);
+
+        _setTimeWindow(timeWindow_);
+    }
+
+    /**
+     * @notice Updates the price data for all the registered Uniswap V2 pairs
+     *
+     * May be called at any time. The time window automatically adjusts
+     */
+    function updatePrices() public virtual {
+        uint256 pairsLength_ = _pairs.length();
+
+        for (uint256 i = 0; i < pairsLength_; i++) {
+            address pair_ = _pairs.at(i);
+
+            PairInfo storage pairInfo = _pairInfos[pair_];
+            uint256[] storage pairTimestamps = pairInfo.blockTimestamps;
+
+            (uint256 price0Cumulative_, uint256 price1Cumulative_, uint256 blockTimestamp_) = pair_
+                .currentCumulativePrices();
+
+            if (
+                pairTimestamps.length == 0 ||
+                blockTimestamp_ > pairTimestamps[pairTimestamps.length - 1]
+            ) {
+                pairInfo.prices0Cumulative.push(price0Cumulative_);
+                pairInfo.prices1Cumulative.push(price1Cumulative_);
+                pairInfo.blockTimestamps.push(blockTimestamp_);
+            }
+        }
+    }
+
+    /**
+     * @notice The function to retrieve the price of a token following the configured route
+     * @param tokenIn_ The input token address
+     * @param amount_ The amount of the input token
+     * @return The price in the last token of the route and the output token address
```

Comment:
It is better to describe each value under a separate tag

Reply:
Yes